runtime/src/iree-amd-aie/driver/xrt/pipeline_layout.cc

```diff
@@ -0,0 +1,192 @@
+  }
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
+
+static void iree_hal_xrt_descriptor_set_layout_destroy(
+    iree_hal_descriptor_set_layout_t* base_descriptor_set_layout) {
+  iree_hal_xrt_descriptor_set_layout_t* descriptor_set_layout =
+      iree_hal_xrt_descriptor_set_layout_cast(base_descriptor_set_layout);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_allocator_free(descriptor_set_layout->host_allocator,
+                      descriptor_set_layout);
+
+  IREE_TRACE_ZONE_END(z0);
+}
+
+iree_hal_descriptor_set_layout_binding_t*
+iree_hal_xrt_descriptor_set_layout_binding(
+    iree_hal_descriptor_set_layout_t* base_descriptor_set_layout,
+    uint32_t binding) {
+  iree_hal_xrt_descriptor_set_layout_t* descriptor_set_layout =
+      iree_hal_xrt_descriptor_set_layout_cast(base_descriptor_set_layout);
+  for (iree_host_size_t i = 0; i < descriptor_set_layout->binding_count; ++i) {
+    if (descriptor_set_layout->bindings[i].binding == binding) {
+      return &descriptor_set_layout->bindings[i];
+    }
+  }
+  return NULL;
+}
+
+namespace {
+const iree_hal_descriptor_set_layout_vtable_t
+    iree_hal_xrt_descriptor_set_layout_vtable = {
+        /*.destroy = */ iree_hal_xrt_descriptor_set_layout_destroy,
+};
+}  // namespace
+
+//===------------------------------------------------------------------------------------------===//
+// iree_hal_xrt_pipeline_layout_t
+//===------------------------------------------------------------------------------------------===//
+
+typedef struct iree_hal_xrt_pipeline_layout_t {
+  // Abstract resource used for injecting reference counting and vtable; must be
+  // at offset 0.
+  iree_hal_resource_t resource;
+
+  iree_allocator_t host_allocator;
+
+  iree_host_size_t push_constant_count;
+
+  iree_host_size_t set_layout_count;
+  iree_hal_descriptor_set_layout_t* set_layouts[];
+} iree_hal_xrt_pipeline_layout_t;
+
+namespace {
+extern const iree_hal_pipeline_layout_vtable_t
+    iree_hal_xrt_pipeline_layout_vtable;
+}  // namespace
+
+static iree_hal_xrt_pipeline_layout_t* iree_hal_xrt_pipeline_layout_cast(
+    iree_hal_pipeline_layout_t* base_value) {
+  IREE_HAL_ASSERT_TYPE(base_value, &iree_hal_xrt_pipeline_layout_vtable);
+  return (iree_hal_xrt_pipeline_layout_t*)base_value;
+}
+
+iree_status_t iree_hal_xrt_pipeline_layout_create(
+    iree_allocator_t host_allocator, iree_host_size_t set_layout_count,
+    iree_hal_descriptor_set_layout_t* const* set_layouts,
+    iree_host_size_t push_constant_count,
+    iree_hal_pipeline_layout_t** out_pipeline_layout) {
+  IREE_ASSERT_ARGUMENT(!set_layout_count || set_layouts);
+  IREE_ASSERT_ARGUMENT(out_pipeline_layout);
+  *out_pipeline_layout = NULL;
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  iree_hal_xrt_pipeline_layout_t* pipeline_layout = NULL;
+  iree_host_size_t total_size =
+      sizeof(*pipeline_layout) +
+      set_layout_count * sizeof(pipeline_layout->set_layouts[0]);
+  iree_status_t status = iree_allocator_malloc(host_allocator, total_size,
```

Comment:
We can use `IREE_RETURN_AND_END_ZONE_IF_ERROR` here to avoid one level of indenting.